CoreCLR versions cleanup

jkotas opened the pull request:
- Reduce number of layers of version macro definitions

AaronRobinsonMSFT commented:
Perhaps a good time to consider https://github.com/dotnet/runtime/issues/13031?

am11 commented:
One other thing that could be considered in the realm of version and flattening stuff by reducing number of handwritten wrappers around the wrappers is [`SOVERSION`](https://cmake.org/cmake/help/latest/prop_tgt/SOVERSION.html) via `set_target_properties()` cmake function, which maps to the SONAME on Unix, the ABI compatibility of shared binary object. For the various components within the coreclr at build time or if someone is building against, e.g. libcoreclr.so.3 (or something more public), the linker will check the binary compatibility bit and gives error (e.g. if libcoreclr.so.4 is present, it will complain that libcoreclr.so.3 not found); at runtime it gets emitted in `NEEDED` under ELF dynamic section, but does not have any side effect during the runtime (it only errors during the build/link time).

iow, something like this: https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/Native/Windows/clrcompression/zlib/CMakeLists.txt#L189

stephentoub commented on src/coreclr/src/vm/frameworkexceptionloader.cpp:
```diff
@@ -16,16 +16,15 @@ struct ExceptionLocationData
+#define DEFINE_EXCEPTION_IN_OTHER_FX_ASSEMBLY(ns, reKind, assemblySimpleName, bHRformessage, ...) { ns, PTR_CSTR((TADDR) # reKind), assemblySimpleName  },
 #include "rexcep.h"
-    {NULL, NULL, NULL, NULL}  // On Silverlight, this table may be empty.  This dummy entry allows us to compile.
+    {NULL, NULL, NULL}  // On Silverlight, this table may be empty.  This dummy entry allows us to compile.
```
Given the comment, is this still relevant?

jkotas commented:
> Perhaps a good time to consider #13031?

I am trying to avoid changing what goes into the native version resources in this PR. In fact, I have done a few tweaks to make them identical. You can distinguish private builds from official builds via file version number today. The private builds have file version 42,42,42,42424 . This is [Arcade feature](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Version.targets). If we want to do something more or different, it should be proposed as change in Arcade.

jkotas commented:
> SOVERSION

We are not using SOVERSIONs today and I do not see what we would gained by using them. We do not expect anybody linking against our native binaries and we manage the versions (of native binaries and everything else) via directory structure.

AaronRobinsonMSFT commented:
>The private builds have file version 42,42,42,42424 . This is Arcade feature. If we want to do something more or different, it should be proposed as change in Arcade.

I don't think we need anything more. I believe you mentioned this in the referenced issue anyways. I will close that issue since there is something we have in place.

jkotas commented:
@elinor-fung @vitek-karas I am deleting `FX_PRODUCT_VERSION` with this change. Do you have any concerns?

jkotas committed:
CoreCLR versions cleanup

- Use current version numbers everywhere
- Use product version (e.g. 5.0.0) for all user facing versions
- Use file version (e.g. 5.0.20.21010) only when the exact version information is required
- Reduce number of layers of version macro definitions

jkotas commented:
I am going to split this into two PRs: One with the CoreCLR cleanup, and second with `FX_PRODUCT_VERSION` cleanup.

jkotas commented:
Test failures are known issue #35438

AndyAyersMS commented:
I think this or something related might have broken incremental rebuilds?

When I touch a jit file I now end up rebuilding a good chunk of the runtime, because runtime_version.h is getting unconditionally written early in the build.

jkotas commented:
@AndyAyersMS Thanks for reporting this. https://github.com/dotnet/runtime/pull/35732  should fix it.

AndyAyersMS commented:
It inspired me to learn how to view the binlogs.

Thanks for the quick fix.
